### PR TITLE
Add hotsos collector

### PR DESCRIPTION
**Pull request by sombrafam**

### CODE REVIEW

File: juju/README.md [hunk 1]
```diff
@@ -45,3 +47,21 @@ your infrastructure.
...
+## UA HotSOS collector
+
+This tool collects hotsos information from a running juju deployment. It logs
+into the available machines, installs and runs the hotsos collect command.
+The results are then collected and stored in a tarball. After the collection,
+the tool will clean up any installed package from the machines.
+
+### Running the tool
+
+By default, the tool will collect hotsos from 1 unit for each application.
```
Comment (pponnuvel) [inline juju/README.md:60]: Is it a random unit of an application? I think the leader unit would be preferable when multiple units exist.

File: juju/hotsos-collector.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+    logger.setLevel(logging.DEBUG)
+
+    return logger
+
+
+LOG = configure_logging()
+
+
+def process_juju_status(status):
+    principle_charm_apps = {}
```
Comment (pponnuvel) [inline juju/hotsos-collector.py:262]: I think you meant to name this *principal*?

**Comment by dosaboy** [comment]
I'm having issues running this:
```
Collecting data from 12 targets: cinder, glance, glance-simplestreams-sync, keystone, mysql, neutron-api, nova-cloud-controller, nova-compute, ovn-central, placement, rabbitmq-server, vault               
Data and logs will be saved in /tmp/hotsos-collection-fji41h6b                                        
Collecting hotsos reports for cinder, machine: 0                                                      
Exception in thread CollectorTask-1:                                                                  
Traceback (most recent call last):                                                                    
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner                               
    self.run()                                                                                        
  File "/usr/lib/python3.8/threading.py", line 870, in run                                                                                                                                                  
    self._target(*self._args, **self._kwargs)                                                         
  File "./juju/hotsos-collector.py", line 379, in worker                                              
    collector.teardown()                                                                              
Collecting hotsos reports for glance, machine: 1                                                      
  File "./juju/hotsos-collector.py", line 163, in teardown                                            
    self._run("sudo apt remove -y hotsos")                                                            
  File "./juju/hotsos-collector.py", line 134, in _run                                                
    out, err = self.executor.run(cmd, self)                                                           
  File "./juju/hotsos-collector.py", line 49, in run                                                  
    parent_caller.print_status(msg=f"Running: {' '.join(ssh_cmd)}")                                   
AttributeError: 'NoneType' object has no attribute 'print_status'  
```
Maybe I am missing some dependencies.

File: juju/hotsos-collector.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+            fmt = "{}: {}".format(self, msg)
+        else:
+            fmt = "{}".format(self)
+        with THREAD_LOCK:
+            LOG.debug(fmt)
+
+    def _run(self, cmd):
+        if args.debug:
+            with open(self.log_file, "a") as f:
+                out, err = self.executor.run(cmd, f, self)
```
Comment (dosaboy) [inline juju/hotsos-collector.py:129]: the self arg seems unnecessary here since the method is owned by self.

File: juju/hotsos-collector.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+        ]
+
+        try:
+            parent_caller.print_status(msg=f"Running: {' '.join(ssh_cmd)}")
+            process = subprocess.Popen(ssh_cmd, stdout=subprocess.PIPE,
+                                       stderr=subprocess.PIPE,
+                                       universal_newlines=True)
+
+            def read_stdout():
+                nonlocal out
```
Comment (dosaboy) [inline juju/hotsos-collector.py:55]: self.out perhaps

File: juju/hotsos-collector.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+            cmd
+        ]
+
+        try:
+            parent_caller.print_status(msg=f"Running: {' '.join(ssh_cmd)}")
+            process = subprocess.Popen(ssh_cmd, stdout=subprocess.PIPE,
+                                       stderr=subprocess.PIPE,
+                                       universal_newlines=True)
+
+            def read_stdout():
```
Comment (dosaboy) [inline juju/hotsos-collector.py:54]: nested methods should be avoided. if needs be create a new class for these methods

**Review by dosaboy — COMMENTED** [review]
If you rebase your branch off main and run tox it'll give you a few pylint/pep8 errors that need resolving.

File: juju/hotsos-collector.py (hunk 2 above)
Comment (sombrafam) [reply juju/hotsos-collector.py:262]: This is used interchangeably in juju world.

File: juju/README.md (hunk 1 above)
Comment (sombrafam) [reply juju/README.md:60]: Done

File: juju/hotsos-collector.py (hunk 3 above)
Comment (sombrafam) [reply juju/hotsos-collector.py:129]: Not sure if I understand what you mean here, _run is a class method and should have self as first parameter so I can access other elements on the object.

**Comment by sombrafam** [comment]
@dosaboy regarding the error, it's a race condition. It happened once with me but was gone when I was trying to reproduce and fix. Can you collect the folder with the logs so I can look?

**Pull request closed without merging**